### DISCUSSION TITLE
[Feature] Support for "Ultra HDR" pictures on Android

### DISCUSSION THREAD

**Discussion by zsolt-donca** [immich-app/immich#7262] [Feature Request]
### The feature

I noticed when I open my photos on my Pixel 7 phone in Google Photos, the images are displayed in an HDR mode. The way it works is that right after opening the photo it starts out as is "bland", but then just after a fraction of a second it "pops out": the whites become whiter, the colors become more saturated, the whole thing becomes much more pleasing to the eye.

 If I check the details of the photo, it states "Ultra HDR":

![image](https://github.com/immich-app/immich/assets/6824167/3ad8b814-80ee-4df3-a8c3-f6dd23262aed)

It would be great if Immich Android app could display these HDR photos in a similar way, with the added brightness and color depth.

Here is a sample photo:

![PXL_20240211_113336435](https://github.com/immich-app/immich/assets/6824167/ebadbccc-f8ea-4077-8b77-085f21e2505e)

### Platform

- [ ] Server
- [ ] Web
- [X] Mobile

**Comment by mertalev** [comment]
I'd like to add this, but it would require [upstream support](https://github.com/libvips/libvips/issues/3799). Something more actionable in the short-term would be to support AVIF, since this format allows for 10-bit HDR.

**Reply by n00mkrad** [reply]
AVIF is irrelevant.

No camera shoots AVIF.

The only relevant HDR formats are JPEG (with UltraHDR gain map) and HEIF (Apple) as well as potentially some raw formats.

**Reply by mertalev** [reply]
The previews and thumbnails we generate can be AVIF. The fact that the format supports 10-bit HDR with much better compression efficiency than JPEG makes it a good candidate for this.

Of course, the other part of this feature request is to have the mobile app actually display the asset in HDR. Flutter brings its own set of challenges there.

**Reply by n00mkrad** [reply]
Is there a spec for AVIF with a gain map?

No need to reinvent the wheel in case it does not.

**Reply by mertalev** [reply]
AVIF officially supports HDR both natively and through a gain map. I think the latter is much less common, though.

**Comment by martabal** [comment]
Seems like ImageMagick supports ultrahdr https://github.com/ImageMagick/ImageMagick/releases/tag/7.1.1-30 thanks to [libultrahdr](https://github.com/google/libultrahdr)

**Reply by n00mkrad** [reply]
Neat but I'm pretty sure immich uses libvips and not IM

**Reply by martabal** [reply]
Libvips uses imagemagick if it does not support the image format. This is why Immich can decode raw images.

**Reply by marrobHD** [reply]
The feature has been introduced in libvips 8.18 last year. However libvips is only for generating images.

**Reply by jcupitt** [reply]
libvips 8.18 added libuhdr and libraw support at the end of last year:

https://www.libvips.org/2025/12/04/What's-new-in-8.18.html#ultrahdr-support

https://www.libvips.org/2025/12/04/What's-new-in-8.18.html#camera-raw-support

So you can now process raw and uhdr images quickly. Among other things, the next libvips adds CICP support, so you can read and write HDR AVIF, JXL and PNG.

**Comment by apparle** [comment]
Do we need `libvips` support just to display the image or only when editing/modifying/resizing it?
If the image is local on my phone and can be displayed in full quality, is there a stop-gap implementation to just display it correctly with proper Ultra HDR brightness by enabling the HDR mode as in https://developer.android.com/media/grow/ultra-hdr/display ? 
All new Android phones are adopting this format, and the pictures look bland when viewed through immich app -- the difference is big enough to be a deal-breaker for layman "why do my photos look worse in this app, I don't like it", and it looks like libvips is going to take its time :-(

**Reply by mertalev** [reply]
libvips is only to generate Ultra HDR preview images. The actual displaying of Ultra HDR in the app is a separate issue that's limited by Flutter. We're switching to a native video player for HDR support to get around Flutter limitations, and might have to do something similar for images.

**Reply by apparle** [reply]
Is there a separate issue to track the displaying of Ultra HDR, which I should follow? 
I'm realizing that I care more about being able to see UltraHDR in the image viewer once I open a stock camera photo, because that's the most common usage. 
Thumbnails and editing is all gravy, we should eventually get there -- but I don't need / won't notice on a daily basis.

**Comment by bigjam3** [comment]
This is a huge deal for me personally, I love my HDR images.

However I’m on iOS, is there a feature request anywhere for iOS and HEIF gain map HDR photos?

**Comment by TheBeaconCrafter** [comment]
Is Ultra HDR on the roadmap? I'd also love to have this feature available as it's a big deal for me.

**Reply by mertalev** [reply]
It is planned, but not a high priority at the moment. Displaying Ultra HDR images likely requires us to write a custom image library backed by native Kotlin/Swift viewers, which may or may not also need us to handle gestures, zooming, swiping between assets, etc. We wouldn't need to if Flutter's rendering engine had better HDR support, but they unfortunately haven't prioritized HDR as much as I'd like.

**Reply by Delivator** [reply]
What about support for the web app? The "full size thumbnail" (not sure what you'd call them), lack ultra HDR support and only show correctly when I start zooming in a little. I assume it loads the original full image when I start zooming in but it would be nice if it would show correctly when I'm viewing photos in fullscreen without having to zoom in for every image.

**Reply by M1cha** [reply]
Wouldn't it be easier to add HDR support to flutter and create a PR instead of rewriting everything in Kotlin and Swift?

**Reply by dunestorm333** [reply]
Just purchased the $100 supporter package to try and bump this on the priority list. HDR photos is the one feature holding me back from using Immich as my default image viewer!

**Reply by mertalev** [reply]
> Wouldn't it be easier to add HDR support to flutter and create a PR instead of rewriting everything in Kotlin and Swift?

No, a PR for that would not fit in a single PR, it would be a milestone that involves a lot of planning and touches many things from their rendering engine to their APIs. It's much more feasible to use the existing native support than for us to contribute this functionality to Flutter.

We've been getting more familiar with Swift/Kotlin development and have a more established process for it now, so it's actually not that crazy to use native viewers. A PoC would be interesting to see how well it would work. Still no promises on when something like this would land, though.

**Comment by n00mkrad** [comment]
@mertalev 
libvips now seems to support UltraHDR: https://github.com/libvips/libvips/pull/4687

This should be enough to at least enable UltraHDR support on Chromium (and any other browser that supports gain maps).

**Reply by jcupitt** [reply]
The umbrella issue for ultrahdr in libvips is here: https://github.com/libvips/libvips/issues/3799

There will be a few more PRs yet, then this should be part of libvips 8.18 before then end of the year (hopefully).

libvips git master has all this stuff in experimental form, so any testing and feedback before then would be great!

**Reply by jcupitt** [reply]
libvips 8.18 was released in December, there's a chapter in the docs introducing the new UltraHDR API:

https://www.libvips.org/API/current/uhdr.html

**Comment by MrMPN** [comment]
I've been playing a bit with this. I used a bit of AI to figure out the Flutter part (register the native view I guess), the rest is trivial as I'm an Android Developer.
But no matter how much I try, there's just no HDR.

I don't know Flutter, although I've learned a bit toying with this. The ultra HDR native samples in Kotlin that show how to display an image are really simple, so I think the issue is in how Flutter uses the view. I understand Flutter renderer has no HDR support, so we need to render outside of it, and then compose the Flutter UI on top (but tbh, that's for later, for now I just want to get to display an Ultra HDR image). Is that a reasonable way to look at this?

Any ideas on how to go about this? I'll keep trying, as I really would like this feature.

**Reply by MrMPN** [reply]
Nevermind, I got ultra HDR displaying properly. That's just the first step but hopefully the rest is easy and I can open a PR for this

**Reply by sznowicki** [reply]
Kudos for taking this one!

**Comment by MrMPN** [comment]
Ok, I'm gonna post about the issues I've found, as I'd like some feedback.

First, implementing a native view to display the image is straightforward. A simple imageView + Glide (plus proper color mode) does the job no problem.
Should this new ImageViewer be the new default for Android? That's what I've done for now. iOS gets the old/regular one.

Using the imageView for just displaying, and nothing else, is the best idea. I tried making the view handle zoom, but once you mix gestures from Flutter and Native Android, it's hell. It's best to let Flutter control everything.

The first issue is image size, there's a limit to how big the image we load can be. I think it depends on the device, but I crashed with OOM on some pictures from the demo. We can do hardware mode (bitmaps on GPU memory), which is generally better. But that's minSDK 26, and I don't know what happens when we ran out of memory, not sure how defensively I can code this.
I know nothing of how this is handled in Flutter, does Immich do downsampling?

Subsampling with tiling is the gold standard (like [telephoto](https://github.com/saket/telephoto) does) I guess, but the amount of headaches it brings for the gestures is huge. I tried but quickly stopped. 

The other issue is caching. Right now, all caching is in Flutter side. The most seamless implementation is to have a flutter placeholder with thumbnail, load the image below, and once is loaded, fade out/in the native viewer. it kinda looks like what Google Photo does when you open one picture (the HDR kicks in gradually for like 300ms), which is nice. But it does so for every picture as you move between them, so unlike Google Photos. It also means that no matter if the thumbnail has gainmap (libvips update helps here?), we would not see it, as it's Flutter side.

There's also a weird flicker when moving from an image to a video. I'll look into it.

What I have is very usable, although there's these 2 (+1) matter that I'm not convinced with. I'll push my branch and link my repo tomorrow in case anyone want to try it out.

**Reply by mertalev** [reply]
I only tried the native view approach on iOS, where getting pan and zoom for original res images worked fine (with some caching inefficiencies as you noted) and felt smooth. The main issue I encountered was that animations for swiping between images and going from timeline to viewer were not smooth and caused a poorer overall UX. We'd have to do something similar to the native video player and show a Flutter-side placeholder image. My idea was to maybe handle the thumbnail and preview on the Flutter side and only use the native viewer for original res images. I didn't get that far with it, though.

The other thing to note is that Flutter's impeller engine [recently landed RGBA32 support](https://github.com/flutter/flutter/pull/177959), so there might actually be hope for HDR without going native. In this particular case, it's kind of more efficient to keep things in Flutter's rendering engine since the platform view is inherently much more expensive to display and mix with Flutter's rendering.

**Comment by griffi-gh** [comment]
This is one of tilhe main things stopping me from using Immich as primary photo/gallery app

Is it planned in the future?
Would it require re-processing all photos or is it purely client-side limitation of the Flutter app?

**Reply by TheBeaconCrafter** [reply]
> This is one of tilhe main things stopping me from using Immich as primary photo/gallery app
> 
> 
> 
> Is it planned in the future?
> 
> Would it require re-processing all photos or is it purely client-side limitation of the Flutter app?

AFAIK this is purely a clientside restriction as Flutter does not support it. If you go into the Immich app and download the photo at full resolution to your phones gallery, the Ultra HDR effect works fine.

Edit: I stand corrected

**Reply by n00mkrad** [reply]
> Is it planned in the future? Would it require re-processing all photos or is it purely client-side limitation of the Flutter app?

It would require re-processing all thumbnails/previews since currently, the gain map is lost during conversion.
Optimally, of course, there could be an option to re-encode **only** UltraHDR previews/thumbs, that would save users who only have, say, 10% of their photos in UltraHDR a lot of time.

**Reply by mertalev** [reply]
> Optimally, of course, there could be an option to re-encode only UltraHDR previews/thumbs, that would save users who only have, say, 10% of their photos in UltraHDR a lot of time.

I think it's unlikely for us to add a specific option like that. Realistically, it will probably mean regenerating all images.

**Comment by FluffyDiscord** [comment]
Would it be possible to consider also HDR PNGs or other formats on mobile and/or browser, than UltraHDR?

**Reply by jcupitt** [reply]
libvips uses Google's libultrahdr for UltraHDR handling and the current version only supports jpg files. Google have started work on a new version of libultrahdr which will be more format-agnostic, but it's not done yet. 

Confusingly, PNGs also support HDR via CICP metadata. There's a libvips branch with CICP support and it'll probably be released in 8.19, the summer 2026 update. 

https://github.com/libvips/libvips/pull/4818

**Reply by n00mkrad** [reply]
This would result in an *insane* increase in bandwidth (and storage) requirements as PNG is lossless.
HDR PNG support would be neat but UltraHDR JPEG is far more important right now as that's a format that comes straight out of a phone camera, and one that has reasonable file sizes.

**Reply by miguelpruivo** [reply]
@n00mkrad agree although I must add “UltraHDR HEIC” as well because most of iOS photos these days are HEIC and not jpeg (and some qualities don’t allow jpeg at all). Also it’s the default photo quality for iOS.

**Comment by miguelpruivo** [comment]
Looking forward this too. The HDR photos on iOS default gallery look stunning, I'd really love to see this implemented on Immich App as well. 

With Flutter, have you considered using [Android](https://docs.flutter.dev/platform-integration/android/platform-views) and [iOS](https://docs-flutter-dev.translate.goog/platform-integration/ios/platform-views?_x_tr_sl=en&_x_tr_tl=pt-PT&_x_tr_hl=pt-PT&_x_tr_pto=sc) native views for showing the media files? Using that, we should be able to use the same native approach as other apps use (ie., Google Photos etc) and getting rid of the Flutter engine limitations in this regard.

From what I read, we can't use `UIImageView` for that and we need to use Metal, but haven't played with it.

You may also want to check [this](https://stackoverflow.com/questions/66024649/how-to-display-an-hdr-photo-like-in-photos-app-that-the-highlight-and-shadow-pa).

**Reply by FluffyDiscord** [reply]
the issue with native views is I assume that you have little to no control over them

**Reply by miguelpruivo** [reply]
You have because you are the one creating the native view (ie., Swift and Kotlin) and then embedding with Flutter.

**Reply by mertalev** [reply]
The main thing is that it causes a stutter when the platform view is made because of how Flutter creates it internally. There is actually a PR for this (#26294) that displays a Flutter-side placeholder like for videos, but it still has some jank going on.

**Reply by miguelpruivo** [reply]
@mertalev on videos though it seems that Immich already has full support for HDR or am I wrong? At least my HDR videos seem to render the same way as in the photos app.

suggestion: it would be “cool” to have a water mark or something when viewing in full screen that would say it’s using HDR, DV etc or not. Even iOS photos app had that before but no longer. 

Edit: found [this app](https://apps.apple.com/pt/app/radiance/id1573366225) that apparently can enhance any non HDR photo into ultra HDR resulting in exactly what’s expected to see in the photos app (and desired for Immich as well). This is probably a native app but this reinforces the fact that with a native approach + optimized platform views we may be able get it.

I’ll check the PR if I have time and give it some try to see if it can be improved further.

**Reply by n00mkrad** [reply]
@miguelpruivo I don't know about videos, but when it comes to photos, gain maps are stripped when the previews are encoded.

Also @mertalev this thread is about UltraHDR pictures with gain maps. HDR Video is completely different tech and off-topic imo.

**Reply by miguelpruivo** [reply]
@n00mkrad yes I was just mentioning because that means that there’s no “limitation” in flutter engine being able to render HDR nit ranges if that’s a concern.

**Comment by mv-l** [comment]
I would also really like to see this feature.

**Comment by bverkron** [comment]
Just chiming in to say that HDR is an absolutely MASSIVE feature for me.

I've been working on / off as a photographer for a couple decades and HDR is one of the biggest improvements to photo viewing I've ever experienced. As such I've moved most of my workflow to be HDR-first and am really struggling to find a platform that's capable of sharing/displaying HDR images properly.

Immich was my fist choice because I was already running it and LOVE it, but I've had to explore running alternative platforms in parallel just for HDR support which is not ideal. Currently I'm running Ente Photo self-hosted for testing and while it does technically work for HDR it's got a _lot_ of shortcomings compared to Immich. Their mobile iOS app doesn't support HDR but their web app does and the web app is clunky and not a great experience for end users.

Here's hoping that Immich can add support for this in the near future. HDR is becoming more and more popular and many modern smartphones shoot HDR by default and it feels like going backwards to have to view modern HDR photos in SDR.

Love this project and thanks for all the efforts!

**Comment by CoderCore** [comment]
For web pages, as long as the correct HDR browser is provided, the content will display correctly. However, for client-side applications, specific settings may be required. Currently, neither browsers nor client-side applications support HDR. The main issue with browsers is that they don't retain the necessary HDR information when generating JPEG thumbnails. This problem should be easily solved by optimizing the transcoding library used. However, for client-side applications, it might be more complex. For professional videographers, HDR is becoming increasingly important. IMMICH is a very good software in many aspects, but it would be a great pity if it didn't support HDR.

**Comment by alcino** [comment]
Can someone help me understand the current status of HDR support?

If I got it right, currently

- It's supported in the browser if you load the original picture
  - The option to load the original picture seems to be ignored unless you pinch zoom?
- It's not supported in the mobile apps
  - This is a Flutter limitation that can't really be addressed here
  - There was some discussion about JPEG XL support here https://github.com/immich-app/immich/discussions/8284 that seems to suggest it was dependent on a libvips PR that has recently been merged: https://github.com/libvips/libvips/pull/4979

Did I miss anything? Is there anything that a non-expert in these frameworks can do to help move this forward?

**Reply by bverkron** [reply]
I would also love to know. No HDR is a show stopper for me for full adoption. One of the devs (I think?) commented on a Reddit thread a few months ago stating it should not be too difficult to add but I cannot find that right now and it’s unclear in what context they meant that.

**Comment by wnagrodzki** [comment]
Would it be possible to split this feature into two smaller steps?

1. Preserve HDR data (when the source images contain it) when generating JPEG thumbnails, previews, and full-size images.
2. Update the mobile clients to handle HDR images correctly.

This would allow the web client to take advantage of HDR, as mentioned in https://github.com/immich-app/immich/discussions/29802.

I recently started exporting photos as HDR JPEGs, and the lack of HDR in previews is especially noticeable compared with the lack of HDR in thumbnails.

**Reply by jcupitt** [reply]
As far as I know, 1. is already done for UltraHDR JPEG. Current stable libvips (8.18) has all the code, all you need to do is update and view in a desktop browser. You'll need to rebuild any cached thumbnails after updating I guess.

The next stable version (8.19), due this autumn, adds support for HDR AVIF, PNG, JXL, and Apple gainmaps, which should help further.

Your 2. is probably more difficult!

**Reply by wnagrodzki** [reply]
@jcupitt Just confirmed this is not true. Preview generation drops gainmap during the process.

**Reply by jcupitt** [reply]
Oh no! I guess immich is explicitly stripping metadata then. It's probably only a one-line change somewhere.

**Reply by jcupitt** [reply]
Oh, immich is using sharp, I'll ask if anything needs doing there.

**Reply by kleisauke** [reply]
I suspect this depends on PR https://github.com/immich-app/immich/pull/29132 and there's probably also a [`.keepGainMap()`](https://sharp.pixelplumbing.com/api-output/#keepgainmap) call needed here:
https://github.com/immich-app/immich/blob/1b8cbaab9762dc53fa9b15366f22b65dc91a5e81/server/src/repositories/media.repository.ts#L187-L195

Also, the patch added in PR https://github.com/immich-app/base-images/pull/360 may need to be revised, since the `uhdrload` priority needs to be bumped as well (the merged patch was based on an earlier version of PR https://github.com/libvips/libvips/pull/5125).

**Reply by mertalev** [reply]
libvips itself is capable of outputting Ultra HDR JPEGs as of 8.18 (which we use), but I'm not sure if we build it with libultrahdr at the moment. The sharp API and build patch are good shouts too - that might also be an issue.

**Reply by n00mkrad** [reply]
I hope this gets a fairly high priority. This is the \#1 thing stopping me from ditching Google Photos. It sucks to shoot nice photos and as soon as you sync them off of your phone, the HDR map is gone.
